vmdb/app/controllers/provider_foreman_controller.rb

```diff
@@ -0,0 +1,535 @@
+    case model
+    when "Provider"
+      @foreman = identify_record(id, ProviderForeman)
+      if @foreman.nil?
+        self.x_node = "root"
+        get_node_info("root")
+        return
+      else
+        options = {:model => "ConfigurationProfileForeman"}
+        options[:where_clause] = ["configuration_manager_id IN (?)", @foreman.id]
+        process_show_list(options)
+        @right_cell_text =
+          _("%{model} \"%{name}\"") %
+          {:name  => @foreman.name,
+           :model => "#{ui_lookup(:model => model ? model : TreeBuilder.get_model_for_prefix(@nodetype))}"}
+      end
+    when "ConfigurationProfile"
+      cpf = identify_record(id, ConfigurationProfileForeman)
+      if cpf.nil?
+        self.x_node = "root"
+        get_node_info("root")
+        return
+      else
+        options = {:model => "ConfiguredSystemForeman"}
+        options[:where_clause] = ["configuration_profile_id IN (?)", cpf.id]
+        process_show_list(options)
+        @right_cell_text =
+          _("%{model} \"%{name}\"") %
+          {:name  => cpf.name,
+           :model => "#{ui_lookup(:model => model ? model : TreeBuilder.get_model_for_prefix(@nodetype))}"}
+      end
+    when "ConfiguredSystem"
+      @record = identify_record(id, ConfiguredSystemForeman) unless @record
+      if @record.nil?
+        self.x_node = "root"
+        get_node_info("root")
+        return
+      else
+        show_record(from_cid(id))
+        @right_cell_text =
+          _("%{model} \"%{name}\"") %
+          {:name  => @record.name,
+           :model => "#{ui_lookup(:model => model ? model : TreeBuilder.get_model_for_prefix(@nodetype))}"}
+      end
+    when "MiqSearch"
+      options = {:model => "ConfiguredSystemForeman"}
+      process_show_list(options)
+      @right_cell_text = _("All Foreman Configured Systems")
+    else      # Get list of child VMs of this node
+      if x_node == "root"
+        if self.x_active_tree == :foreman_providers_tree
+          options = {:model => "ProviderForeman"}
+          process_show_list(options)
+          @right_cell_text = _("All Foreman Providers")
+        elsif self.x_active_tree == :cs_filter_tree
+          options = {:model => "ConfiguredSystemForeman"}
+          process_show_list(options)
+          @right_cell_text = _("All Foreman Configured Systems")
+        end
+      end
+    end
+    @right_cell_text += @edit[:adv_search_applied][:text] if x_tree[:type] == :filter && @edit && @edit[:adv_search_applied]
+
+    if @edit && @edit.fetch_path(:adv_search_applied, :qs_exp) # If qs is active, save it in history
+      x_history_add_item(:id     => x_node,
+                         :qs_exp => @edit[:adv_search_applied][:qs_exp],
+                         :text   => @right_cell_text)
+    else
+      x_history_add_item(:id => treenodeid, :text => @right_cell_text)  # Add to history pulldown array
+    end
+  end
+
+  def replace_right_cell(replace_trees = [])
```

Comment:
There are no extra methods in this controller.
All we need to do is split huge chunks of code that is in one method into smaller methods, perhaps even have separate subclasses for handling tree related and list related operations.
I agree that this being a new controller we should not repeat past mistakes. However, I think Phase 1 should be to just get things working.